operators/pkg/instautoctrl/expiration.go

```diff
@@ -0,0 +1,290 @@
+		}
+		return ctrl.Result{}, err
+	}
+
+	// check if the instance reached the maximum time of lifetime and if so delete it
+	isDeleted, err := r.deleteStaleInstance(ctx, &instance)
+	if err != nil {
+		log.Error(err, "failed delete-stale-instance")
+	}
+	if isDeleted {
+		log.Info("Instance has been deleted", "instance", instance.Name)
+		tenant, err := r.GetTenantFromInstance(ctx, &instance)
+		if err != nil {
+			log.Error(err, "failed retrieving tenant from instance")
+			return ctrl.Result{}, err
+		}
+		err = r.SendNotification(ctx, &instance, tenant.Spec.Email)
+		if err != nil {
+			log.Error(err, "failed sending notification email")
+			return ctrl.Result{}, err
+		}
+		log.Info("Notification email sent to user", "instance", instance.Name, "email", tenant.Spec.Email)
+		tracer.Step("Stale instance deleted")
+		return ctrl.Result{}, nil
+	}
+
+	tracer.Step("stale instance done")
+
+	dbgLog.Info("requeueing instance")
+	return ctrl.Result{RequeueAfter: 24 * time.Hour}, nil // TODO revisit this value, it is a placeholder.
+}
+
+// TerminateInstance terminates the Instance.
+// TODO move somewhere else, as it is used in other controllers too.
+func (r *InstanceExpirationReconciler) TerminateInstance(ctx context.Context, instance *clv1alpha2.Instance) error {
+	log := ctrl.LoggerFrom(ctx).WithName("termination")
+	log.Info("Terminating instance", "instance", instance.Name, " in namespace", instance.Namespace)
+
+	var template clv1alpha2.Template
+	var err = r.Get(ctx, types.NamespacedName{
+		Name:      instance.Spec.Template.Name,
+		Namespace: instance.Namespace,
+	}, &template)
+	if err != nil {
+		log.Error(err, "Unable to fetch the instance template.")
+		return err
+	}
+
+	var environment = template.Spec.EnvironmentList[0]
+	if environment.Persistent {
+		log.Info("Stopping persistent instance...")
+		instance.Spec.Running = false
+		return r.Update(ctx, instance)
+	}
+	log.Info("Deleting non-persistent instance...")
+	return r.Delete(ctx, instance)
+}
+
+// SendNotification sends an email to the user to notify that the instance will be terminated/stopped if they do not use it anymore.
+func (r *InstanceExpirationReconciler) SendNotification(ctx context.Context, instance *clv1alpha2.Instance, userEmail string) error {
+	log := ctrl.LoggerFrom(ctx).WithName("notification-email-instance")
+	log.Info("sending email notification to user", "instance", instance.Name, "email", userEmail)
+	emailBody := fmt.Sprintf(
+		"Dear user,\n\n"+
+			"Your instance %s has reached the maximum lifetime and has now been terminated.\n\n"+
+			"Best regards,\n"+
+			"CrownLabs Team",
+		instance.Name,
+	)
+	err := r.MailClient.SendMail([]string{userEmail}, "CrownLabs Instance Termination Alert", emailBody)
+	if err != nil {
+		log.Error(err, "failed sending email notification")
+		return err
+	}
+	log.Info("The notification to the tenant has been sent", "instance", instance.Name)
+
+	return nil
+}
```

Comment:
I'd make this more generic:
you can have an initial email common header and a footer, then a message that can possibly change. If you make a function that receives the instance, the tenant and the variable message (possibly adding a simple templating system in which you substituite {name} and {prettyName} (the latter is the one that the user sees on the dashboard) I think it can also be recycled elsewhere.